### PR TITLE
BLAKE2 Compression Function F Precompile 

### PR DESCRIPTION
**Pull request by fedejinich**
Goals of this PR
- Merge [this other PR](https://github.com/rsksmart/rskj/pull/1049) into master
- Be compatible with [EIP-152](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-152.md)

I'm treating this new precompile as blackbox: byte array goes in byte array goes out

Also a [little fix](https://github.com/rsksmart/rskj/commit/454570093065ea8b9a6603f2dae41f3a4ae5a578) has been made regarding `ExtcodeHash` OPCODE tests.

NOTE: RSKIP176 will be added soon :)

### CODE REVIEW

File: rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java [hunk 1]
```diff
@@ -228,4 +233,103 @@ public void modExpTest() {
...
+        DataWord addr = DataWord.valueFromHex("0000000000000000000000000000000000000000000000000000000000000009");
+        PrecompiledContract contract = precompiledContracts.getContractForAddress(activations, addr);
+        assertNotNull(contract);
+
+        byte[] dataBytes = Hex.decode(data);
+
+        try {
+            byte[] resultBytes = contract.execute(dataBytes);
+
+            assertEquals(expected, Hex.toHexString(resultBytes));
```
Comment (patogallaiovlabs) [inline rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java:319]: Be careful here.
I would leave the error assertion outside or in a new method, and add a "fail()" if it doesnt fail.
Could happen, that asserts / doesnt fail, and it will be still green when it shouldnt.

File: rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java [hunk 2]
```diff
@@ -228,4 +233,103 @@ public void modExpTest() {
...
+    public void blake2fTestVector8() {
+        // This tests takes a lot of time to be executed
+
+        String data = "ffffffff48c9bdf267e6096a3ba7ca8485ae67bb2bf894fe72f36e3cf1361d5f3af54fa5d182e6ad7f520e511f6c3e2b8c68059b6bbd41fbabd9831f79217e1319cde05b61626300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000300000000000000000000000000000001";
+        String expected = "fc59093aafa9ab43daae0e914c57635c5402d8e3d2130eb9b3cc181de7f0ecf9b22bf99a7815ce16419e200e01846e6b5df8cc7703041bbceb571de6631d2615";
+
+        internalBlake2fTestVector(data, expected, null);
+    }
+
+    private void internalBlake2fTestVector(String data, String expected, String expectedError) {
```
Comment (patogallaiovlabs) [inline rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java:306]: shouldnt we assert also for gas consumed? not sure where.

**Review by patogallaiovlabs — CHANGES_REQUESTED** [review]

File: rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java (hunk 1 above)
Comment (fedejinich) [reply rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java]: modified!

File: rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java (hunk 2 above)
Comment (fedejinich) [reply rskj-core/src/test/java/org/ethereum/vm/PrecompiledContractTest.java]: https://github.com/rsksmart/rskj/pull/1314/commits/afb1b587b639700fba7dfb94df30c24426f856ed

**Comment by fedejinich** [comment]
`testVector8` covers a really weird edge case but it delays test execution by 3 minutes, I've marked it as Ignored

**Comment by fedejinich** [comment]
I had to adapt some rskip174 tests

**Comment by fedejinich** [comment]
pipeline:run

**Comment by fedejinich** [comment]
need to rename rskip176 to a new one

**Comment by patogallaiovlabs** [comment]
pipeline:run

File: rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
+package org.ethereum.crypto.cryptohash;
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertArrayEquals;
+
+/**
+ * Created by bakaking on 25/10/2019.
+ */
+public class Black2bTest {
```
Comment (ajlopezrsk) [inline rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java:10]: wrong class name?

File: rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+        boolean f = true;
+        int rounds = 12;
+
+        long[] expected = {
+                0x0D4D1C983FA580BAL, 0xE9F6129FB697276AL,
+                0xB7C45A68142F214CL, 0xD1A2FFDB6FBB124BL,
+                0x2D79AB2A39C5877DL, 0x95CC3345DED552C2L,
+                0x5A92F1DBA88AD318L, 0x239900D4ED8623B9L};
+
+        Blake2b.functionF(h, m, c, f, rounds);
```
Comment (ajlopezrsk) [inline rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java:38]: Any chance to have more tests, with different input arguments, like number of rownds or f in false

**Review by ajlopezrsk — CHANGES_REQUESTED** [review]

File: rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java (hunk 3 above)
Comment (fedejinich) [reply rskj-core/src/test/java/org/ethereum/crypto/cryptohash/Black2bTest.java]: https://github.com/rsksmart/rskj/pull/1314/commits/fd1d7b33933dbf9a779a2c772948b29c7388a74a

**Review by donequis — APPROVED** [review]

**Review by ajlopezrsk — APPROVED** [review]
